### ISSUE TITLE
React Native: Observables and PubSub

### ISSUE THREAD

**Issue by zefre** [aws-amplify/amplify-js#750]
## Do you want to request a feature or report a bug?
This appears to be a bug.

## What is the current behavior?
When subscribing to a topic as suggested in the Amplify docs, react native gives the error: TypeError: One of the sources for assign has an enumerable key on the prototype chain. Are you trying to assign to a prototype property? 

## If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. Your bug will get fixed much faster if we can run your code and it doesn't have dependencies other than AWS Amplify.
```js
import Amplify, { PubSub } from 'aws-amplify';
import { AWSIoTProvider } from 'aws-amplify/lib/PubSub/Providers';
Amplify.addPuggable(new AWSIoTProvider({
        aws_pubsub_region: "VALID_REGION",
        aws_pubsub_endpoint: "VALID_ENDPOINT",
      })
);
PubSub.subscribe('someTopicLiteral').subscribe((msg) => console.log(msg));
// or
PubSub.subscribe('someTopicLiteral2').subscribe({next: (msg) => console.log(msg)});
```

## What is the expected behavior?
PubSub would subscribe to the topic with the specified listener.

## Which versions of Amplify, and which browser / OS are affected by this issue? Did this work in previous versions?
iOS React Native, with Amplify ^0.2.11. I don't believe PubSub existed in previous versions

You can turn on the debug mode to provide more info for us by setting ```window.LOG_LEVEL = 'DEBUG';``` in your app.

**Comment by richardlevano** [comment]
Same here I am using  
"dependencies": {
    "aws-amplify": "^0.3.3",
    "aws-amplify-react": "^0.1.43",
    "aws-amplify-react-native": "^0.2.9",
    "react": "16.3.1",
    "react-native": "0.55.3"
  },

and 
PubSub.subscribe('myTopicTest').subscribe({
      next: data => console.log('Message received', data),
      error: error => console.error(error),
      close: () => console.log('Done'),
    });
Throws that error.
Can not figure out a way around it  :(.

**Comment by johnrjj** [comment]
Having same issue over here.Happens as early as the first function call

`PubSub.subscribe('myTopic');`
/*
[Unhandled promise rejection: TypeError: One of the sources for assign has an enumerable key on the prototype chain. Are you trying to assign a prototype property? We don't allow it, as this is an edge case that we do not support. This error is a performance optimization and not spec compliant.]
*/

**Comment by johnrjj** [comment]
Found the source of the error
```
    var canonicalRequest = canonical_request({
        method: 'GET',
        url: url.format(__assign({}, parsedUrl, { query: __assign({}, parsedUrl.query, queryParams) })),
        headers: signedHeaders,
    });
```
inside `signUrl` in `Signer.js`

parsedQuery has things on the prototype that are being improperly assigned (from the 'url' npm package)

**Comment by richardlevano** [comment]
Any update om this issue. This is a stopper for PubSub aws-amplify use.

**Comment by johnrjj** [comment]
If you want a hacky workaround in the meantime (I needed one), go into`/node_modules/aws-amplify/lib/Common/Signer.js` 

and around line 248 change
```
 var canonicalRequest = canonical_request({
        method: 'GET',
        url: url.format(__assign({}, parsedUrl, { query: __assign({}, parsedUrl.query, queryParams) })), // this line
        headers: signedHeaders,
    });
```
to 
```
var canonicalRequest = canonical_request({
        method: 'GET',
        url: url.format(__assign(parsedUrl, { query: __assign({}, parsedUrl.query, queryParams) })),
        headers: signedHeaders,
    });
```

**Comment by richardlevano** [comment]
Now after 

    PubSub.subscribe('myTopicTest').subscribe({
            next: data => console.log('Message received', data),
            error: error => console.error(error),
        close: () => console.log('Done'),
      });

I get

   Possible Unhandled Promise Rejection (id: 0):
    Object {
        "errorCode": 1,
       "errorMessage": "AMQJSC0001E Connect timed out.",
       "invocationContext": undefined,
    }

Any hints?

**Comment by johnrjj** [comment]
Yeah, I ran into that right after moving forward. Going to take a leap and say it's the same issue: 

Assuming you ran the 
`aws iot attach-principal-policy --policy-name 'myIOTPolicy' --principal '<YOUR_COGNITO_IDENTITY_ID>'` to attach the policy to IoT Core,
you also then have to connect your cognito group to an IoT policy (they left this step out of the docs).

For testing sake, if you add the full `AWSIoTFullAccess` policy to your Cognito user pool you're using and try to connect, it should work. I'm still thinking about out how to write a good policy for the user pool in the meantime.

**Comment by richardlevano** [comment]
Have to connect a cognito group to an IoT policy?

**Comment by richardlevano** [comment]
I have finally receiving messages from  mqtt subscribe but I had to modify the myIOTPolicy to 
    {
     "Version": "2012-10-17",
     "Statement": [
       {
         "Effect": "Allow",
         "Action": "iot:*",
         "Resource": "*"
       }
     ]
    }

Thanks johnrjj

**Comment by richardlevano** [comment]
dennis-b,

Besides the creation of a policy in aws IOT Core which is attached to a "Thing". The same or another policy with similar permissions should be attach to your cognito pool user too in order to get the mqtt messages.

For the test purposes try 
{
"Action": "iot:*"
...
...
}
in your policy.

**Comment by dennis-b** [comment]
@richardlevano thanks man, it works working.
but now i see that the messages is not perssist, when my mobile phone is off line then the messages wis not delivered, ahh i will try to create new ticket on this one

**Comment by richardlevano** [comment]
In my case the mqtt message also stopped due to the disconnection of the secure websocket (wss).
So you will probably have to try to keep the connection alive.....perhaps a keepalive publish on regular intervals.

**Comment by dennis-b** [comment]
eventually i implemented my one MqttOverWSProvider with paho.mqtt.javascript and recconect that looks like this : 
`
 client.onConnected = (reconnect, uri) => {
      if (reconnect) {
        targetTopics.forEach(topic => { client.subscribe(topic) })
       }
    }
`
it  still not supporting offline mode, but it works after reconnect now

**Comment by kevex91** [comment]
This is still an issue on the current version of aws-amplify. Had to apply the small fix to Signer.js otherwise TypeScript complains.

**Comment by paulmand3l** [comment]
Just ran into this issue. Any progress?

**Comment by elorzafe** [comment]
Fix is already published on aws-amplify@unstable (1.1.4-unstable.2)
